File: packages/auth/src/rbac.ts
type UserRole =
  | 'SENIOR'
  | 'FAMILY'
  | 'CAREGIVER_APPLICANT'
  | 'CAREGIVER_APPROVED'
  | 'PROVIDER_ADMIN'
  | 'PROVIDER_STAFF'
  | 'MEDICAL_STAFF'
  | 'GOVERNMENT_REVIEWER'
  | 'PARTNER_OPERATOR'
  | 'PLATFORM_ADMIN';

// ---------------------------------------------------------------------------
// Enums
// ---------------------------------------------------------------------------

export enum Resource {
  PROFILE = 'PROFILE',
  SENIOR_PROFILE = 'SENIOR_PROFILE',
  CARE_PLAN = 'CARE_PLAN',
  VISIT = 'VISIT',
  MEDICATION = 'MEDICATION',
  MEDICAL_HISTORY = 'MEDICAL_HISTORY',
  MATCH_REQUEST = 'MATCH_REQUEST',
  PROVIDER = 'PROVIDER',
  CAREGIVER_APPLICATION = 'CAREGIVER_APPLICATION',
  ELIGIBILITY_CASE = 'ELIGIBILITY_CASE',
  OBSERVABILITY = 'OBSERVABILITY',
  NOTIFICATION = 'NOTIFICATION',
  CONSENT = 'CONSENT',
  AUDIT_LOG = 'AUDIT_LOG',
  REFERRAL = 'REFERRAL',
  APPOINTMENT = 'APPOINTMENT',
}

export enum Action {
  CREATE = 'CREATE',
  READ = 'READ',
  UPDATE = 'UPDATE',
  DELETE = 'DELETE',
  LIST = 'LIST',
  APPROVE = 'APPROVE',
  REJECT = 'REJECT',
}

export enum Scope {
  OWN = 'OWN',
  ASSIGNED = 'ASSIGNED',
  ORG = 'ORG',
  ALL = 'ALL',
}

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface Permission {
  resource: Resource;
  action: Action;
  scope: Scope;
}

// ---------------------------------------------------------------------------
// Role → Permission mappings
// ---------------------------------------------------------------------------

const p = (resource: Resource, action: Action, scope: Scope): Permission => ({
  resource,
  action,
  scope,
});

export const ROLE_PERMISSIONS: Record<UserRole, Permission[]> = {
  SENIOR: [
    // Own profile
    p(Resource.PROFILE, Action.READ, Scope.OWN),
    p(Resource.PROFILE, Action.UPDATE, Scope.OWN),
    p(Resource.SENIOR_PROFILE, Action.READ, Scope.OWN),
    p(Resource.SENIOR_PROFILE, Action.UPDATE, Scope.OWN),
    // Own medications
    p(Resource.MEDICATION, Action.READ, Scope.OWN),
    p(Resource.MEDICATION, Action.LIST, Scope.OWN),
    // Own medical history (read-only)
    p(Resource.MEDICAL_HISTORY, Action.READ, Scope.OWN),
    p(Resource.MEDICAL_HISTORY, Action.LIST, Scope.OWN),
    // Care plan (read-only)
    p(Resource.CARE_PLAN, Action.READ, Scope.OWN),
    p(Resource.CARE_PLAN, Action.LIST, Scope.OWN),
    // Consent management
    p(Resource.CONSENT, Action.CREATE, Scope.OWN),
    p(Resource.CONSENT, Action.READ, Scope.OWN),
    p(Resource.CONSENT, Action.UPDATE, Scope.OWN),
    p(Resource.CONSENT, Action.DELETE, Scope.OWN),
    p(Resource.CONSENT, Action.LIST, Scope.OWN),
    // Appointments
    p(Resource.APPOINTMENT, Action.CREATE, Scope.OWN),
    p(Resource.APPOINTMENT, Action.READ, Scope.OWN),
    p(Resource.APPOINTMENT, Action.UPDATE, Scope.OWN),
    p(Resource.APPOINTMENT, Action.LIST, Scope.OWN),
    // Notifications
    p(Resource.NOTIFICATION, Action.READ, Scope.OWN),
    p(Resource.NOTIFICATION, Action.LIST, Scope.OWN),
    p(Resource.NOTIFICATION, Action.UPDATE, Scope.OWN),
  ],

  FAMILY: [
    // Own profile
    p(Resource.PROFILE, Action.READ, Scope.OWN),
    p(Resource.PROFILE, Action.UPDATE, Scope.OWN),
    // Delegated senior data (assigned = linked senior)
    p(Resource.SENIOR_PROFILE, Action.READ, Scope.ASSIGNED),
    p(Resource.MEDICATION, Action.READ, Scope.ASSIGNED),
    p(Resource.MEDICATION, Action.LIST, Scope.ASSIGNED),
    p(Resource.MEDICAL_HISTORY, Action.READ, Scope.ASSIGNED),
    p(Resource.MEDICAL_HISTORY, Action.LIST, Scope.ASSIGNED),
    p(Resource.CARE_PLAN, Action.READ, Scope.ASSIGNED),
    p(Resource.CARE_PLAN, Action.LIST, Scope.ASSIGNED),
    p(Resource.VISIT, Action.READ, Scope.ASSIGNED),
    p(Resource.VISIT, Action.LIST, Scope.ASSIGNED),
    // Match requests on behalf of senior
    p(Resource.MATCH_REQUEST, Action.CREATE, Scope.ASSIGNED),
    p(Resource.MATCH_REQUEST, Action.READ, Scope.ASSIGNED),
    p(Resource.MATCH_REQUEST, Action.UPDATE, Scope.ASSIGNED),
    p(Resource.MATCH_REQUEST, Action.LIST, Scope.ASSIGNED),
    // Consent approval on behalf of senior
    p(Resource.CONSENT, Action.APPROVE, Scope.ASSIGNED),
    p(Resource.CONSENT, Action.READ, Scope.ASSIGNED),
    p(Resource.CONSENT, Action.LIST, Scope.ASSIGNED),
    // Appointments
    p(Resource.APPOINTMENT, Action.CREATE, Scope.ASSIGNED),
    p(Resource.APPOINTMENT, Action.READ, Scope.ASSIGNED),
    p(Resource.APPOINTMENT, Action.UPDATE, Scope.ASSIGNED),
    p(Resource.APPOINTMENT, Action.LIST, Scope.ASSIGNED),
    // Notifications
    p(Resource.NOTIFICATION, Action.READ, Scope.OWN),
    p(Resource.NOTIFICATION, Action.LIST, Scope.OWN),
    p(Resource.NOTIFICATION, Action.UPDATE, Scope.OWN),
  ],

  CAREGIVER_APPLICANT: [
    // Own profile
    p(Resource.PROFILE, Action.READ, Scope.OWN),
    p(Resource.PROFILE, Action.UPDATE, Scope.OWN),
    // Own caregiver application and credentials
    p(Resource.CAREGIVER_APPLICATION, Action.CREATE, Scope.OWN),
    p(Resource.CAREGIVER_APPLICATION, Action.READ, Scope.OWN),
    p(Resource.CAREGIVER_APPLICATION, Action.UPDATE, Scope.OWN),
    p(Resource.CAREGIVER_APPLICATION, Action.LIST, Scope.OWN),
    // Notifications
    p(Resource.NOTIFICATION, Action.READ, Scope.OWN),
    p(Resource.NOTIFICATION, Action.LIST, Scope.OWN),
    p(Resource.NOTIFICATION, Action.UPDATE, Scope.OWN),
  ],

  CAREGIVER_APPROVED: [
    // Own profile
    p(Resource.PROFILE, Action.READ, Scope.OWN),
    p(Resource.PROFILE, Action.UPDATE, Scope.OWN),
    // Own application
    p(Resource.CAREGIVER_APPLICATION, Action.READ, Scope.OWN),
    p(Resource.CAREGIVER_APPLICATION, Action.UPDATE, Scope.OWN),
    // Assigned clients
    p(Resource.SENIOR_PROFILE, Action.READ, Scope.ASSIGNED),
    p(Resource.CARE_PLAN, Action.READ, Scope.ASSIGNED),
    p(Resource.CARE_PLAN, Action.LIST, Scope.ASSIGNED),
    // Visits
    p(Resource.VISIT, Action.READ, Scope.ASSIGNED),
    p(Resource.VISIT, Action.UPDATE, Scope.ASSIGNED),
    p(Resource.VISIT, Action.LIST, Scope.ASSIGNED),
    // Medications (assigned clients, read-only + event recording)
    p(Resource.MEDICATION, Action.READ, Scope.ASSIGNED),
    p(Resource.MEDICATION, Action.LIST, Scope.ASSIGNED),
    p(Resource.MEDICATION, Action.UPDATE, Scope.ASSIGNED),
    // Observability
    p(Resource.OBSERVABILITY, Action.CREATE, Scope.ASSIGNED),
    p(Resource.OBSERVABILITY, Action.READ, Scope.ASSIGNED),
    // Notifications
    p(Resource.NOTIFICATION, Action.READ, Scope.OWN),
    p(Resource.NOTIFICATION, Action.LIST, Scope.OWN),
    p(Resource.NOTIFICATION, Action.UPDATE, Scope.OWN),
  ],

  PROVIDER_ADMIN: [
    // Own org profile management
    p(Resource.PROVIDER, Action.READ, Scope.OWN),
    p(Resource.PROVIDER, Action.UPDATE, Scope.OWN),
    // Org caregivers
    p(Resource.CAREGIVER_APPLICATION, Action.READ, Scope.ORG),
    p(Resource.CAREGIVER_APPLICATION, Action.LIST, Scope.ORG),
    p(Resource.CAREGIVER_APPLICATION, Action.APPROVE, Scope.ORG),
    p(Resource.CAREGIVER_APPLICATION, Action.REJECT, Scope.ORG),
    p(Resource.CAREGIVER_APPLICATION, Action.UPDATE, Scope.ORG),
    // Org clients and senior profiles
    p(Resource.SENIOR_PROFILE, Action.READ, Scope.ORG),
    p(Resource.SENIOR_PROFILE, Action.LIST, Scope.ORG),
    p(Resource.PROFILE, Action.READ, Scope.ORG),
    p(Resource.PROFILE, Action.LIST, Scope.ORG),
    // Care plans
    p(Resource.CARE_PLAN, Action.CREATE, Scope.ORG),
    p(Resource.CARE_PLAN, Action.READ, Scope.ORG),
    p(Resource.CARE_PLAN, Action.UPDATE, Scope.ORG),
    p(Resource.CARE_PLAN, Action.DELETE, Scope.ORG),
    p(Resource.CARE_PLAN, Action.LIST, Scope.ORG),
    // Visits
    p(Resource.VISIT, Action.CREATE, Scope.ORG),
    p(Resource.VISIT, Action.READ, Scope.ORG),
    p(Resource.VISIT, Action.UPDATE, Scope.ORG),
    p(Resource.VISIT, Action.LIST, Scope.ORG),
    // Referrals
    p(Resource.REFERRAL, Action.CREATE, Scope.ORG),
    p(Resource.REFERRAL, Action.READ, Scope.ORG),
    p(Resource.REFERRAL, Action.UPDATE, Scope.ORG),
    p(Resource.REFERRAL, Action.LIST, Scope.ORG),
    p(Resource.REFERRAL, Action.APPROVE, Scope.ORG),
    // Observability
    p(Resource.OBSERVABILITY, Action.READ, Scope.ORG),
    p(Resource.OBSERVABILITY, Action.LIST, Scope.ORG),
    // Notifications
    p(Resource.NOTIFICATION, Action.READ, Scope.OWN),
    p(Resource.NOTIFICATION, Action.LIST, Scope.OWN),
    p(Resource.NOTIFICATION, Action.UPDATE, Scope.OWN),
    p(Resource.NOTIFICATION, Action.CREATE, Scope.ORG),
  ],

  PROVIDER_STAFF: [
    // Own profile
    p(Resource.PROFILE, Action.READ, Scope.OWN),
    p(Resource.PROFILE, Action.UPDATE, Scope.OWN),
    // Assigned clients
    p(Resource.SENIOR_PROFILE, Action.READ, Scope.ASSIGNED),
    p(Resource.PROFILE, Action.READ, Scope.ASSIGNED),
    // Care plans (assigned)
    p(Resource.CARE_PLAN, Action.READ, Scope.ASSIGNED),
    p(Resource.CARE_PLAN, Action.UPDATE, Scope.ASSIGNED),
    p(Resource.CARE_PLAN, Action.LIST, Scope.ASSIGNED),
    // Visits (assigned)
    p(Resource.VISIT, Action.CREATE, Scope.ASSIGNED),
    p(Resource.VISIT, Action.READ, Scope.ASSIGNED),
    p(Resource.VISIT, Action.UPDATE, Scope.ASSIGNED),
    p(Resource.VISIT, Action.LIST, Scope.ASSIGNED),
    // Referrals (read assigned)
    p(Resource.REFERRAL, Action.READ, Scope.ASSIGNED),
    p(Resource.REFERRAL, Action.LIST, Scope.ASSIGNED),
    // Observability
    p(Resource.OBSERVABILITY, Action.CREATE, Scope.ASSIGNED),
    p(Resource.OBSERVABILITY, Action.READ, Scope.ASSIGNED),
    // Notifications
    p(Resource.NOTIFICATION, Action.READ, Scope.OWN),
    p(Resource.NOTIFICATION, Action.LIST, Scope.OWN),
    p(Resource.NOTIFICATION, Action.UPDATE, Scope.OWN),
  ],

  MEDICAL_STAFF: [
    // Own profile
    p(Resource.PROFILE, Action.READ, Scope.OWN),
    p(Resource.PROFILE, Action.UPDATE, Scope.OWN),
    // Medical history (with consent)
    p(Resource.MEDICAL_HISTORY, Action.CREATE, Scope.ASSIGNED),
    p(Resource.MEDICAL_HISTORY, Action.READ, Scope.ASSIGNED),
    p(Resource.MEDICAL_HISTORY, Action.UPDATE, Scope.ASSIGNED),
    p(Resource.MEDICAL_HISTORY, Action.LIST, Scope.ASSIGNED),
    // Medications (assigned)
    p(Resource.MEDICATION, Action.CREATE, Scope.ASSIGNED),
    p(Resource.MEDICATION, Action.READ, Scope.ASSIGNED),
    p(Resource.MEDICATION, Action.UPDATE, Scope.ASSIGNED),
    p(Resource.MEDICATION, Action.LIST, Scope.ASSIGNED),
    // Referrals
    p(Resource.REFERRAL, Action.CREATE, Scope.ASSIGNED),
    p(Resource.REFERRAL, Action.READ, Scope.ASSIGNED),
    p(Resource.REFERRAL, Action.UPDATE, Scope.ASSIGNED),
    p(Resource.REFERRAL, Action.LIST, Scope.ASSIGNED),
    // Appointments
    p(Resource.APPOINTMENT, Action.CREATE, Scope.ASSIGNED),
    p(Resource.APPOINTMENT, Action.READ, Scope.ASSIGNED),
    p(Resource.APPOINTMENT, Action.UPDATE, Scope.ASSIGNED),
    p(Resource.APPOINTMENT, Action.LIST, Scope.ASSIGNED),
    // Consent (read to verify access)
    p(Resource.CONSENT, Action.READ, Scope.ASSIGNED),
    // Notifications
    p(Resource.NOTIFICATION, Action.READ, Scope.OWN),
    p(Resource.NOTIFICATION, Action.LIST, Scope.OWN),
    p(Resource.NOTIFICATION, Action.UPDATE, Scope.OWN),
  ],

  GOVERNMENT_REVIEWER: [
    // Read-only on eligibility cases
    p(Resource.ELIGIBILITY_CASE, Action.READ, Scope.ALL),
    p(Resource.ELIGIBILITY_CASE, Action.LIST, Scope.ALL),
    p(Resource.ELIGIBILITY_CASE, Action.UPDATE, Scope.ALL),
    p(Resource.ELIGIBILITY_CASE, Action.APPROVE, Scope.ALL),
    p(Resource.ELIGIBILITY_CASE, Action.REJECT, Scope.ALL),
    // Provider registry
    p(Resource.PROVIDER, Action.READ, Scope.ALL),
    p(Resource.PROVIDER, Action.LIST, Scope.ALL),
    // Observability (read-only)
    p(Resource.OBSERVABILITY, Action.READ, Scope.ALL),
    p(Resource.OBSERVABILITY, Action.LIST, Scope.ALL),
    // Audit logs (read-only)
    p(Resource.AUDIT_LOG, Action.READ, Scope.ALL),
    p(Resource.AUDIT_LOG, Action.LIST, Scope.ALL),
    // Notifications
    p(Resource.NOTIFICATION, Action.READ, Scope.OWN),
    p(Resource.NOTIFICATION, Action.LIST, Scope.OWN),
    p(Resource.NOTIFICATION, Action.UPDATE, Scope.OWN),
  ],

  PARTNER_OPERATOR: [
    // Own provider profile
    p(Resource.PROVIDER, Action.READ, Scope.OWN),
    p(Resource.PROVIDER, Action.UPDATE, Scope.OWN),
    // Own services
    p(Resource.CARE_PLAN, Action.READ, Scope.OWN),
    p(Resource.CARE_PLAN, Action.LIST, Scope.OWN),
    p(Resource.REFERRAL, Action.READ, Scope.OWN),
    p(Resource.REFERRAL, Action.LIST, Scope.OWN),
    // Notifications
    p(Resource.NOTIFICATION, Action.READ, Scope.OWN),
    p(Resource.NOTIFICATION, Action.LIST, Scope.OWN),
    p(Resource.NOTIFICATION, Action.UPDATE, Scope.OWN),
  ],

  PLATFORM_ADMIN: [
    // Full access to all resources
    ...Object.values(Resource).flatMap((resource) =>
      Object.values(Action).map((action) => p(resource, action, Scope.ALL)),
    ),
  ],
};

// ---------------------------------------------------------------------------
// Helper functions
// ---------------------------------------------------------------------------

/**
 * Returns true if the given role has a permission matching the resource,
 * action, and optional scope. When scope is omitted, any scope matches.
 * PLATFORM_ADMIN always returns true.
 */
export function hasPermission(
  role: UserRole,
  resource: Resource,
  action: Action,
  scope?: Scope,
): boolean {
  const permissions = ROLE_PERMISSIONS[role] ?? [];
  return permissions.some(
    (perm) =>
      perm.resource === resource &&
      perm.action === action &&
      (scope === undefined || perm.scope === scope || perm.scope === Scope.ALL),
  );
}

/**
 * Returns all permissions for the given role.
 */
export function getAllPermissions(role: UserRole): Permission[] {
  return ROLE_PERMISSIONS[role] ?? [];
}

File: crates/types/src/rbac.rs
// =============================================================================
// RBAC — 10 roles × 16 resources × 7 actions × 4 scopes
// Ported from packages/auth/src/rbac.ts (360 lines)
// =============================================================================

use once_cell::sync::Lazy;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use strum::{Display, EnumIter, EnumString, IntoEnumIterator};

use crate::enums::UserRole;

// ---------------------------------------------------------------------------
// Enums
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize, Display, EnumString, EnumIter)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
#[strum(serialize_all = "SCREAMING_SNAKE_CASE")]
pub enum Resource {
    Profile,
    SeniorProfile,
    CarePlan,
    Visit,
    Medication,
    MedicalHistory,
    MatchRequest,
    Provider,
    CaregiverApplication,
    EligibilityCase,
    Observability,
    Notification,
    Consent,
    AuditLog,
    Referral,
    Appointment,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize, Display, EnumString, EnumIter)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
#[strum(serialize_all = "SCREAMING_SNAKE_CASE")]
pub enum Action {
    Create,
    Read,
    Update,
    Delete,
    List,
    Approve,
    Reject,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize, Display, EnumString)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
#[strum(serialize_all = "SCREAMING_SNAKE_CASE")]
pub enum Scope {
    Own,
    Assigned,
    Org,
    All,
}

// ---------------------------------------------------------------------------
// Permission
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Permission {
    pub resource: Resource,
    pub action: Action,
    pub scope: Scope,
}

const fn p(resource: Resource, action: Action, scope: Scope) -> Permission {
    Permission {
        resource,
        action,
        scope,
    }
}

// ---------------------------------------------------------------------------
// Role → Permission mappings
// ---------------------------------------------------------------------------

static ROLE_PERMISSIONS: Lazy<HashMap<UserRole, Vec<Permission>>> = Lazy::new(|| {
    use Action::*;
    use Resource::*;
    use Scope::*;

    let mut map = HashMap::new();

    // --- SENIOR ---
    map.insert(
        UserRole::Senior,
        vec![
            p(Profile, Read, Own),
            p(Profile, Update, Own),
            p(SeniorProfile, Read, Own),
            p(SeniorProfile, Update, Own),
            p(Medication, Read, Own),
            p(Medication, List, Own),
            p(MedicalHistory, Read, Own),
            p(MedicalHistory, List, Own),
            p(CarePlan, Read, Own),
            p(CarePlan, List, Own),
            p(Consent, Create, Own),
            p(Consent, Read, Own),
            p(Consent, Update, Own),
            p(Consent, Delete, Own),
            p(Consent, List, Own),
            p(Appointment, Create, Own),
            p(Appointment, Read, Own),
            p(Appointment, Update, Own),
            p(Appointment, List, Own),
            p(Notification, Read, Own),
            p(Notification, List, Own),
            p(Notification, Update, Own),
        ],
    );

    // --- FAMILY ---
    map.insert(
        UserRole::Family,
        vec![
            p(Profile, Read, Own),
            p(Profile, Update, Own),
            p(SeniorProfile, Read, Assigned),
            p(Medication, Read, Assigned),
            p(Medication, List, Assigned),
            p(MedicalHistory, Read, Assigned),
            p(MedicalHistory, List, Assigned),
            p(CarePlan, Read, Assigned),
            p(CarePlan, List, Assigned),
            p(Visit, Read, Assigned),
            p(Visit, List, Assigned),
            p(MatchRequest, Create, Assigned),
            p(MatchRequest, Read, Assigned),
            p(MatchRequest, Update, Assigned),
            p(MatchRequest, List, Assigned),
            p(Consent, Approve, Assigned),
            p(Consent, Read, Assigned),
            p(Consent, List, Assigned),
            p(Appointment, Create, Assigned),
            p(Appointment, Read, Assigned),
            p(Appointment, Update, Assigned),
            p(Appointment, List, Assigned),
            p(Notification, Read, Own),
            p(Notification, List, Own),
            p(Notification, Update, Own),
        ],
    );

    // --- CAREGIVER_APPLICANT ---
    map.insert(
        UserRole::CaregiverApplicant,
        vec![
            p(Profile, Read, Own),
            p(Profile, Update, Own),
            p(CaregiverApplication, Create, Own),
            p(CaregiverApplication, Read, Own),
            p(CaregiverApplication, Update, Own),
            p(CaregiverApplication, List, Own),
            p(Notification, Read, Own),
            p(Notification, List, Own),
            p(Notification, Update, Own),
        ],
    );

    // --- CAREGIVER_APPROVED ---
    map.insert(
        UserRole::CaregiverApproved,
        vec![
            p(Profile, Read, Own),
            p(Profile, Update, Own),
            p(CaregiverApplication, Read, Own),
            p(CaregiverApplication, Update, Own),
            p(SeniorProfile, Read, Assigned),
            p(CarePlan, Read, Assigned),
            p(CarePlan, List, Assigned),
            p(Visit, Read, Assigned),
            p(Visit, Update, Assigned),
            p(Visit, List, Assigned),
            p(Medication, Read, Assigned),
            p(Medication, List, Assigned),
            p(Medication, Update, Assigned),
            p(Observability, Create, Assigned),
            p(Observability, Read, Assigned),
            p(Notification, Read, Own),
            p(Notification, List, Own),
            p(Notification, Update, Own),
        ],
    );

    // --- PROVIDER_ADMIN ---
    map.insert(
        UserRole::ProviderAdmin,
        vec![
            p(Provider, Read, Own),
            p(Provider, Update, Own),
            p(CaregiverApplication, Read, Org),
            p(CaregiverApplication, List, Org),
            p(CaregiverApplication, Approve, Org),
            p(CaregiverApplication, Reject, Org),
            p(CaregiverApplication, Update, Org),
            p(SeniorProfile, Read, Org),
            p(SeniorProfile, List, Org),
            p(Profile, Read, Org),
            p(Profile, List, Org),
            p(CarePlan, Create, Org),
            p(CarePlan, Read, Org),
            p(CarePlan, Update, Org),
            p(CarePlan, Delete, Org),
            p(CarePlan, List, Org),
            p(Visit, Create, Org),
            p(Visit, Read, Org),
            p(Visit, Update, Org),
            p(Visit, List, Org),
            p(Referral, Create, Org),
            p(Referral, Read, Org),
            p(Referral, Update, Org),
            p(Referral, List, Org),
            p(Referral, Approve, Org),
            p(Observability, Read, Org),
            p(Observability, List, Org),
            p(Notification, Read, Own),
            p(Notification, List, Own),
            p(Notification, Update, Own),
            p(Notification, Create, Org),
        ],
    );

    // --- PROVIDER_STAFF ---
    map.insert(
        UserRole::ProviderStaff,
        vec![
            p(Profile, Read, Own),
            p(Profile, Update, Own),
            p(SeniorProfile, Read, Assigned),
            p(Profile, Read, Assigned),
            p(CarePlan, Read, Assigned),
            p(CarePlan, Update, Assigned),
            p(CarePlan, List, Assigned),
            p(Visit, Create, Assigned),
            p(Visit, Read, Assigned),
            p(Visit, Update, Assigned),
            p(Visit, List, Assigned),
            p(Referral, Read, Assigned),
            p(Referral, List, Assigned),
            p(Observability, Create, Assigned),
            p(Observability, Read, Assigned),
            p(Notification, Read, Own),
            p(Notification, List, Own),
            p(Notification, Update, Own),
        ],
    );

    // --- MEDICAL_STAFF ---
    map.insert(
        UserRole::MedicalStaff,
        vec![
            p(Profile, Read, Own),
            p(Profile, Update, Own),
            p(MedicalHistory, Create, Assigned),
            p(MedicalHistory, Read, Assigned),
            p(MedicalHistory, Update, Assigned),
            p(MedicalHistory, List, Assigned),
            p(Medication, Create, Assigned),
            p(Medication, Read, Assigned),
            p(Medication, Update, Assigned),
            p(Medication, List, Assigned),
            p(Referral, Create, Assigned),
            p(Referral, Read, Assigned),
            p(Referral, Update, Assigned),
            p(Referral, List, Assigned),
            p(Appointment, Create, Assigned),
            p(Appointment, Read, Assigned),
            p(Appointment, Update, Assigned),
            p(Appointment, List, Assigned),
            p(Consent, Read, Assigned),
            p(Notification, Read, Own),
            p(Notification, List, Own),
            p(Notification, Update, Own),
        ],
    );

    // --- GOVERNMENT_REVIEWER ---
    map.insert(
        UserRole::GovernmentReviewer,
        vec![
            p(EligibilityCase, Read, All),
            p(EligibilityCase, List, All),
            p(EligibilityCase, Update, All),
            p(EligibilityCase, Approve, All),
            p(EligibilityCase, Reject, All),
            p(Provider, Read, All),
            p(Provider, List, All),
            p(Observability, Read, All),
            p(Observability, List, All),
            p(AuditLog, Read, All),
            p(AuditLog, List, All),
            p(Notification, Read, Own),
            p(Notification, List, Own),
            p(Notification, Update, Own),
        ],
    );

    // --- PARTNER_OPERATOR ---
    map.insert(
        UserRole::PartnerOperator,
        vec![
            p(Provider, Read, Own),
            p(Provider, Update, Own),
            p(CarePlan, Read, Own),
            p(CarePlan, List, Own),
            p(Referral, Read, Own),
            p(Referral, List, Own),
            p(Notification, Read, Own),
            p(Notification, List, Own),
            p(Notification, Update, Own),
        ],
    );

    // --- PLATFORM_ADMIN — full access ---
    let admin_perms: Vec<Permission> = Resource::iter()
        .flat_map(|resource| Action::iter().map(move |action| p(resource, action, All)))
        .collect();
    map.insert(UserRole::PlatformAdmin, admin_perms);

    map
});

// ---------------------------------------------------------------------------
// Helper functions
// ---------------------------------------------------------------------------

/// Returns true if the given role has a permission matching the resource,
/// action, and optional scope. When scope is None, any scope matches.
/// PLATFORM_ADMIN always returns true.
pub fn has_permission(
    role: UserRole,
    resource: Resource,
    action: Action,
    scope: Option<Scope>,
) -> bool {
    let permissions = match ROLE_PERMISSIONS.get(&role) {
        Some(p) => p,
        None => return false,
    };
    permissions.iter().any(|perm| {
        perm.resource == resource
            && perm.action == action
            && match scope {
                None => true,
                Some(s) => perm.scope == s || perm.scope == Scope::All,
            }
    })
}

/// Returns all permissions for the given role.
pub fn all_permissions(role: UserRole) -> &'static [Permission] {
    ROLE_PERMISSIONS
        .get(&role)
        .map_or(&[], |v| v.as_slice())
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn senior_can_read_own_profile() {
        assert!(has_permission(
            UserRole::Senior,
            Resource::Profile,
            Action::Read,
            Some(Scope::Own),
        ));
    }

    #[test]
    fn senior_cannot_delete_care_plan() {
        assert!(!has_permission(
            UserRole::Senior,
            Resource::CarePlan,
            Action::Delete,
            None,
        ));
    }

    #[test]
    fn family_can_create_match_request_assigned() {
        assert!(has_permission(
            UserRole::Family,
            Resource::MatchRequest,
            Action::Create,
            Some(Scope::Assigned),
        ));
    }

    #[test]
    fn family_cannot_create_match_request_all() {
        assert!(!has_permission(
            UserRole::Family,
            Resource::MatchRequest,
            Action::Create,
            Some(Scope::All),
        ));
    }

    #[test]
    fn provider_admin_can_approve_applications() {
        assert!(has_permission(
            UserRole::ProviderAdmin,
            Resource::CaregiverApplication,
            Action::Approve,
            Some(Scope::Org),
        ));
    }

    #[test]
    fn government_reviewer_can_read_all_eligibility() {
        assert!(has_permission(
            UserRole::GovernmentReviewer,
            Resource::EligibilityCase,
            Action::Read,
            Some(Scope::All),
        ));
    }

    #[test]
    fn government_reviewer_cannot_create_care_plan() {
        assert!(!has_permission(
            UserRole::GovernmentReviewer,
            Resource::CarePlan,
            Action::Create,
            None,
        ));
    }

    #[test]
    fn platform_admin_full_access() {
        // Admin can do anything on any resource
        assert!(has_permission(
            UserRole::PlatformAdmin,
            Resource::AuditLog,
            Action::Delete,
            Some(Scope::All),
        ));
        assert!(has_permission(
            UserRole::PlatformAdmin,
            Resource::CarePlan,
            Action::Create,
            Some(Scope::All),
        ));
    }

    #[test]
    fn caregiver_applicant_limited_access() {
        assert!(has_permission(
            UserRole::CaregiverApplicant,
            Resource::CaregiverApplication,
            Action::Create,
            Some(Scope::Own),
        ));
        assert!(!has_permission(
            UserRole::CaregiverApplicant,
            Resource::CarePlan,
            Action::Read,
            None,
        ));
    }

    #[test]
    fn scope_none_matches_any() {
        // When scope is None, any scope should match
        assert!(has_permission(
            UserRole::Senior,
            Resource::Profile,
            Action::Read,
            None,
        ));
    }
}
